### PR TITLE
Updating permissions so that CryptoOutputStream works

### LINKED ISSUE #76734: HDFS Repository fails when over-the-wire encryption is enabled
**Elasticsearch version** (`bin/elasticsearch --version`): 7.11.1

**Plugins installed**: [repository-hdfs]

**JVM version** (`8 or 15`):

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:
If HDFS is configured to use over-the-wire encryption (i.e. dfs.encrypt.data.transfer is set to true), then the repository-hdfs plugin fails when creating a snapshot. The reason appears to be that this setting results in the HDFS client using CryptoOutputStream, which requires some additional runtime permissions that this plugin is not already granted.
To make matters worse, the version of CryptoOutputStream in the hadoop 2 client is not compatible with java 9 and later. It looks like this has only been addressed in the hadoop 3 client. Here is a stack trace on java 8:

> java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessClassInPackage.sun.nio.ch")
> at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472) ~[?:1.8.0_275]
> at java.security.AccessController.checkPermission(AccessController.java:886) ~[?:1.8.0_275]
> at java.lang.SecurityManager.checkPermission(SecurityManager.java:549) ~[?:1.8.0_275]
> at java.lang.SecurityManager.checkPackageAccess(SecurityManager.java:1564) ~[?:1.8.0_275]
> at java.net.FactoryURLClassLoader.loadClass(URLClassLoader.java:814) ~[?:1.8.0_275]
> at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_275]
> at org.apache.hadoop.crypto.CryptoStreamUtils.freeDB(CryptoStreamUtils.java:39) ~[?:?]
> at org.apache.hadoop.crypto.CryptoInputStream.freeBuffers(CryptoInputStream.java:683) ~[?:?]
> at org.apache.hadoop.crypto.CryptoInputStream.close(CryptoInputStream.java:317) ~[?:?]
> at java.io.FilterInputStream.close(FilterInputStream.java:181) ~[?:1.8.0_275]
> at org.apache.hadoop.hdfs.DataStreamer.closeStream(DataStreamer.java:996) ~[?:?]
> at org.apache.hadoop.hdfs.DataStreamer.closeInternal(DataStreamer.java:839) ~[?:?]
> at org.apache.hadoop.hdfs.DataStreamer.run(DataStreamer.java:834) ~[?:?]

### PR DESCRIPTION
**Pull request by masseyke**
From what I can tell, CryptoOutputStream fails right now because we don't run it with adequate permissions to access the private sun classes that it uses. And it fails on java 11 because in addition to those permissions, the hadoop library uses a class that doesn't exist in Java 9 and later.

### CODE REVIEW

**Comment by masseyke** [comment]
Some possible workarounds to avoid needing this PR (I haven't tested any of these):
- Write your snapshot to an unencrypted zone in HDFS
- Use Java 8. Make the changes locally to plugin-security.policy. And run HA HDFS (looks like we restrict the permissions in code if you're not running in HA mode?)

File: plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsSecurityContext.java [hunk 1]
```diff
@@ -67,10 +67,13 @@
...
             new PrivateCredentialPermission("org.apache.hadoop.security.Credentials * \"*\"", "read"),
             new PrivateCredentialPermission("javax.security.auth.kerberos.KerberosTicket * \"*\"", "read"),
-            new PrivateCredentialPermission("javax.security.auth.kerberos.KeyTab * \"*\"", "read")
+            new PrivateCredentialPermission("javax.security.auth.kerberos.KeyTab * \"*\"", "read"),
             // Included later:
             // 7) allow code to initiate kerberos connections as the logged in user
             // Still far and away fewer permissions than the original full plugin policy
+            // The following are required by CryptoOutputStream:
+            new RuntimePermission("accessClassInPackage.sun.nio.ch"),
+            new RuntimePermission("accessClassInPackage.sun.misc")
```
Comment (jbaiera) [inline plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsSecurityContext.java:76]: Nit: Can we order these above the `// Included later` line and make sure the ordering of these matches up with the ordering with the policy file?

**Review by jbaiera — COMMENTED** [review]
I think this is definitely a permission restriction issue. I think if we're interested in fixing this up, we need to get encrypted data transfer enabled for our secure hdfs tests to ensure that there aren't any other hidden permission surprises underneath the currently breaking call site.

**Comment by masseyke** [comment]
Closing this because it looks like using hadoop 3 client libraries is probably the way to go (since this PR only works on java 8)

**Pull request closed without merging**